docs/en/week13/13-3.md

```diff
@@ -0,0 +1,518 @@
+
+Inside the `forward` function, having called `g.update_all`, we obtain the results of graph convolution `h` and `e`, which represent the terms $(\boldsymbol{Ax} + \sum_{v_j→v}{\eta(\boldsymbol{e_{j}^{h}})\odot \boldsymbol{Bx_{j}}})$ from (Eq.5) and $\boldsymbol{e_{j}}$ from (Eq. 7) respectively. Then, we normalize `h` and `e` with respect to the graph node size and graph edge size respectively. Batch normalization is then applied so that we can train the network effectively. Finally, we apply $\texttt{ReLU}(\cdot)$ and add the residual connections to obtain the hidden representations for the nodes and edges, which are then returned by the `forward` function.
+
+
+```python
+class GatedGCN_layer(nn.Module):
+    
+    def __init__(self, input_dim, output_dim):
+        super().__init__()
+        self.A = nn.Linear(input_dim, output_dim)
+        self.B = nn.Linear(input_dim, output_dim)
+        self.C = nn.Linear(input_dim, output_dim)
+        self.D = nn.Linear(input_dim, output_dim)
+        self.E = nn.Linear(input_dim, output_dim)
+        self.bn_node_h = nn.BatchNorm1d(output_dim)
+        self.bn_node_e = nn.BatchNorm1d(output_dim)
+
+    def message_func(self, edges):
+        Bh_j = edges.src['Bh']
+        # e_ij = Ce_ij + Dhi + Ehj
+        e_ij = edges.data['Ce'] + edges.src['Dh'] + edges.dst['Eh']
+        edges.data['e'] = e_ij
+        return {'Bh_j' : Bh_j, 'e_ij' : e_ij}
+
+    def reduce_func(self, nodes):
+        Ah_i = nodes.data['Ah']
+        Bh_j = nodes.mailbox['Bh_j']
+        e = nodes.mailbox['e_ij']
+        # sigma_ij = sigmoid(e_ij)
+        sigma_ij = torch.sigmoid(e)
+        # hi = Ahi + sum_j eta_ij * Bhj
+        h = Ah_i + torch.sum(sigma_ij * Bh_j, dim=1) / torch.sum(sigma_ij, dim=1)
+        return {'h' : h}
+    
+    def forward(self, g, h, e, snorm_n, snorm_e):
+        
+        h_in = h # residual connection
+        e_in = e # residual connection
+        
+        g.ndata['h']  = h 
+        g.ndata['Ah'] = self.A(h) 
+        g.ndata['Bh'] = self.B(h) 
+        g.ndata['Dh'] = self.D(h)
+        g.ndata['Eh'] = self.E(h) 
+        g.edata['e']  = e 
+        g.edata['Ce'] = self.C(e)
+        
+        g.update_all(self.message_func, self.reduce_func)
+        
+        h = g.ndata['h'] # result of graph convolution
+        e = g.edata['e'] # result of graph convolution
+        
+        h = h * snorm_n # normalize activation w.r.t. graph node size
+        e = e * snorm_e # normalize activation w.r.t. graph edge size
+        
+        h = self.bn_node_h(h) # batch normalization  
+        e = self.bn_node_e(e) # batch normalization  
+        
+        h = torch.relu(h) # non-linear activation
+        e = torch.relu(e) # non-linear activation
+        
+        h = h_in + h # residual connection
+        e = e_in + e # residual connection
+        
+        return h, e
+```
```

Comment:
Break down each method, and explain it separately.

Reply:
Done